File: src/main/java/io/github/binaryfoo/decoders/bit/FullByteField.java
package io.github.binaryfoo.decoders.bit;

import io.github.binaryfoo.bit.EmvBit;

import java.util.HashSet;
import java.util.Set;
import java.util.TreeSet;

public class FullByteField implements BitStringField {

    private final Set<EmvBit> field;
    private final int byteNumber;
    private final String hexValue;
    private final String decodedValue;

    public FullByteField(Set<EmvBit> field, int byteNumber, String hexValue, String decodedValue) {
        this.byteNumber = byteNumber;
        this.hexValue = hexValue;
        this.field = new TreeSet<>(field);
        this.decodedValue = decodedValue;
    }

    @Override
    public String getPositionIn(Set<EmvBit> bitString) {
        return String.format("Byte %d = 0x%s", byteNumber, hexValue);
    }

    @Override
    public String getValueIn(Set<EmvBit> bitString) {
        if (intersects(field, bitString)) {
            return decodedValue;
        }
        return null;
    }

    @Override
    public int getStartBytesOffset() {
        return byteNumber - 1;
    }

    @Override
    public int getLengthInBytes() {
        return 1;
    }

    private boolean intersects(Set<EmvBit> targetBits, Set<EmvBit> bits) {
        Set<EmvBit> intersection = new HashSet<>(targetBits);
        intersection.retainAll(bits);
        return intersection.size() == targetBits.size();
    }

}

File: src/main/java/io/github/binaryfoo/decoders/bit/FullByteField.kt
package io.github.binaryfoo.decoders.bit

import io.github.binaryfoo.bit.EmvBit
import io.github.binaryfoo.bit.matches
import java.util.*

/**
 * An english description for a hex literal of a single byte.
 */
public class FullByteField(field: Set<EmvBit>, private val byteNumber: Int, private val hexValue: String, private val decodedValue: String) : BitStringField {

    private val field: Set<EmvBit>

    init {
        this.field = TreeSet(field)
    }

    override public fun getPositionIn(bits: Set<EmvBit>?): String {
        return "Byte $byteNumber = 0x$hexValue"
    }

    override public fun getValueIn(bits: Set<EmvBit>): String? {
        return if (field.matches(bits)) decodedValue else null
    }

    override fun getStartBytesOffset(): Int = byteNumber - 1

    override fun getLengthInBytes(): Int = 1

}
